homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,281 @@
+    _attr_device_class: ValveDeviceClass | None
+    _attr_is_closed: bool | None = None
+    _attr_is_closing: bool | None = None
+    _attr_is_opening: bool | None = None
+    _attr_supported_features: ValveEntityFeature = ValveEntityFeature(0)
+
+    _valve_is_last_toggle_direction_open = True
+    _attr_reports_position: bool
+
+    @property
+    def reports_position(self) -> bool:
+        """Return True if entity reports position, False otherwise."""
+        if hasattr(self, "_attr_reports_position"):
+            return self._attr_reports_position
+        if hasattr(self, "entity_description"):
+            return self.entity_description.reports_position
+        raise ValueError(f"'reports_position' not set for {self.entity_id}.")
+
+    @property
+    def current_valve_position(self) -> int | None:
+        """Return current position of valve.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._attr_current_valve_position
+
+    @property
+    def device_class(self) -> ValveDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the state of the valve."""
+        reports_position = self.reports_position
+        if self.is_opening:
+            self._valve_is_last_toggle_direction_open = True
+            return STATE_OPENING
+        if self.is_closing:
+            self._valve_is_last_toggle_direction_open = False
+            return STATE_CLOSING
+        if reports_position is True:
+            if (current_valve_position := self.current_valve_position) is None:
+                return None
+            closed: bool | None = current_valve_position == 0
+            return STATE_CLOSED if closed else STATE_OPEN
+        if (closed := self.is_closed) is None:
+            return None
+        return STATE_CLOSED if self.is_closed else STATE_OPEN
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+
+        return {ATTR_CURRENT_POSITION: self.current_valve_position}
+
+    @property
+    def supported_features(self) -> ValveEntityFeature:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def is_opening(self) -> bool | None:
+        """Return if the valve is opening or not."""
+        return self._attr_is_opening
+
+    @property
+    def is_closing(self) -> bool | None:
+        """Return if the valve is closing or not."""
+        return self._attr_is_closing
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return if the valve is closed or not."""
+        if self.reports_position:
```

Comment:
This should just return `self._attr_is_closed`.

Reply:
I believe this is correct.
On previous feedback, it was discussed that `_attr_is_closed` would only be of use on valves that do not report position. On those that do report position, `is_closed` is computed solely based on that position, and the only method that needs to be implemented by people extending this would be `set_valve_position`.

Reply:
Is it possible that someone accesses `is_closed` without accessing `state` first?  If so they would be out of sync, which is what feels weird about setting those properties inside a getter/property like `state`.

Reply:
How would they be out of sync?

Reply:
as in, a valve is open (`_attr_is_closed` is false), then calls `set_valve_position(0)` which sets the valve position to `0` and nothing else (which means `_attr_is_closed`  is still false). And then calls `is_closed` without ever accessing the `state`property before, that call would return `false` as if the valve was still open.

It's not until the user accesses `valve.state` that the value of `_attr_is_closed` gets in sync with the current position of the valve.

Reply:
`ValveEntity.state` doesn't update any state attributes such as `_attr_is_closed`. I'm not following.

Reply:
The only thing we store when the `state` property is called is the last direction in `_valve_is_last_toggle_direction_open`.